modules/transaction-payment/src/lib.rs

```diff
@@ -578,59 +642,199 @@ where
-						let supply_currency_id = *trading_path.first().expect("these's first guaranteed by match");
-						// calculate the supply limit according to oracle price and the slippage limit,
-						// if oracle price is not avalible, do not limit
-						let max_supply_limit = if let Some(target_price) =
-							T::PriceSource::get_relative_price(*target_currency_id, supply_currency_id)
-						{
-							Ratio::one()
-								.saturating_sub(T::MaxSwapSlippageCompareToOracle::get())
-								.reciprocal()
-								.unwrap_or_else(Ratio::max_value)
-								.saturating_mul_int(target_price.saturating_mul_int(amount))
-						} else {
-							PalletBalanceOf::<T>::max_value()
-						};
+		// native is not enough, try swap native from charge fee pool to pay fee and gap.
+		// make sure add extra gap to keep alive after swap.
+		let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
+		Self::swap_native_asset(who, amount);
+	}
+
+	/// Iterate order list, if can swap out enough native asset amount with foreign asset, then stop
+	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
+		let native_currency_id = T::NativeCurrencyId::get();
+		for trading_path in Self::get_trading_path(who) {
+			match trading_path.last() {
+				Some(target_currency_id) if *target_currency_id == native_currency_id => {
+					let supply_currency_id = *trading_path.first().expect("these's first guaranteed by match");
+					if Self::swap_from_treasury(who, amount, supply_currency_id).is_ok() {
+						break;
+					}
+				}
+				_ => {}
+			}
+		}
+	}
+
+	/// swap from treasury
+	#[transactional]
+	fn swap_from_treasury(
+		who: &T::AccountId,
+		amount: Balance,
+		supply_currency_id: CurrencyId,
+	) -> sp_std::result::Result<(), DispatchError> {
+		let treasury_account = Self::treasury_sub_account_id(supply_currency_id);
+		let treasury_supply_balance = T::MultiCurrency::free_balance(supply_currency_id, &treasury_account);
+		let mut rate = FeeRateOfToken::<T>::get(supply_currency_id)
+			.ok_or(Error::<T>::AssetNotInFeePool)
+			.unwrap();
+		let swap_balance_threshold = SwapBalanceThreshold::<T>::get(supply_currency_id);
+
+		// user's supply asset should have enough balance based on old fix rate.
+		let user_supply_balance = T::MultiCurrency::free_balance(supply_currency_id, who);
+		let expect_rate_balance = rate.saturating_mul_int(amount);
+		ensure!(
+			user_supply_balance >= expect_rate_balance,
+			Error::<T>::SupplyBalanceTooLow
+		);
+
+		// if treasury account has not enough native asset, trigger swap
+		let treasury_balance = T::Currency::free_balance(&treasury_account);
+		if treasury_balance < swap_balance_threshold {
+			let trading_path = Self::get_trading_path_by_currency(&treasury_account, supply_currency_id);
+			if let Some(trading_path) = trading_path {
+				Self::deposit_event(Event::TriggerSwap(
+					treasury_balance,
+					swap_balance_threshold,
+					supply_currency_id,
+				));
```

Comment:
`TriggerSwap` here doesn't make sense to me, it's more like a debugging log.

Reply:
ok, sounds too much events.